src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Program.cs

```
@@ -125,23 +125,34 @@ static async Task RunAsync(ServerConfiguration serverConfiguration, Cancellation
 
     var languageServerLogger = loggerFactory.CreateLogger(nameof(LanguageServerHost));
 
-    var (clientPipeName, serverPipeName) = serverConfiguration.ServerPipeName is null
-        ? CreateNewPipeNames()
-        : (serverConfiguration.ServerPipeName, serverConfiguration.ServerPipeName);
+    LanguageServerHost? server = null;
+    if (serverConfiguration.UseStdIo)
+    {
+        server = new LanguageServerHost(Console.OpenStandardInput(), Console.OpenStandardOutput(), exportProvider, languageServerLogger, typeRefResolver);
+        // Redirect Console.Out to try prevent the standard output stream from being corrupted.
+        Console.SetOut(new StreamWriter(Console.OpenStandardError()));
```

We should move this redirection to the top of RunAsync.

---

On this point, should it be just after the logger is created?

```csharp
static async Task RunAsync(ServerConfiguration serverConfiguration, CancellationToken cancellationToken)
{
    // Create a console logger as a fallback to use before the LSP server starts.
    using var loggerFactory = LoggerFactory.Create(builder => ...);

    var logger = loggerFactory.CreateLogger<Program>();

    // Handle LSP communication channel method and Console.Out redirection
    if (serverConfiguration.UseStdIo)
    {
        if (serverConfiguration.ServerPipeName is not null)
        {
            logger.LogError("Server cannot be started with both --stdio and --pipe options.");
            Environment.Exit(1);
        }

        // Redirect Console.Out to try prevent the standard output stream from being corrupted.
        Console.SetOut(new StreamWriter(Console.OpenStandardError()));
    }
    
    ...
```

---

@HarleyRossetto I suspect we actually need to do this before we setup the logger - currently we have a log wrapper that preferentially logs to LSP, but will fallback to logging via stdout if LSP isn't yet initialized.

If we don't redirect stdout before we create the logger, it's possible for the logger to write messages to stdout while LSP initialization is on-going (and possibly corrupt the LSP stream).

```
logger.LogError("Server cannot be started with both --stdio and --pipe options."); Environment.Exit(1);
```

For this I would throw an exception with a custom message so you don't need to utilize the logger

---

@dibarbet Thanks for the further guidance! I've updated my latest commit based on your comment.

I've just thrown a regular exception, is that OK or do you have a preference for another type? 
I did find a [CommandLineArgumentCombinationException](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/CommandLineArgumentCombinationException.cs) type as part of the NuGet.Common package the LSP picks up from a dependency, but I wasn't convinced you would necessarily want to use a custom exception from a dependency.